### PR TITLE
add nx blog

### PR DESCRIPTION
**Pull request by DimiDeKerf**

### CODE REVIEW

File: _posts/2019-12-02-Nx.md [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+You get a set of requirements and start implementing your features.
+After setting up your repository and all the tooling, you’re ready to go.
+All goes well during the development process and you’ve come to the end to deliver your hard work.
+After you’ve deployed your application on the platform, you’re introduced with a blank screen and a console log full of errors.
+
+How can this happen?
+Your team followed the requirements and all the contracts that were defined to communicate with several services.
+Maybe these contracts got outdated in the meantime without you knowing it?
+Were there breaking changes introduced?
+Or does the libraries you are using locally differs from the ones used in production and isn’t compatible anymore?
```
Comment (orjandesmet) [inline _posts/2019-12-02-Nx.md]:
```suggestion
Or does a library you are using locally differ from the ones used in production and isn’t compatible anymore?
```

File: _posts/2019-12-02-Nx.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+But how can you start setting up a monorepo and what tools are required to do so?
+Having to manage all this code in one repo can be really tedious to set up.
+This is where Nx comes into play.
+
+# Nx
+
+[Nx](https://nx.dev/) provides you with the tools to help you develop into a monorepo.
+Taking their experience, the creators of Nx set up the best practices to structure your repository in a maintainable way.
+Nx is built on top of the Angular CLI, offering you a way to create workspaces and scaffolding out of your apps.
```
Comment (orjandesmet) [inline _posts/2019-12-02-Nx.md]: I would mention that they also support React apps in their repository and that you can even configure your own builders to manage apps from other frameworks

File: _posts/2019-12-02-Nx.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+Inside of your CI, you would like to compare the changes between your branch and master.
+This can be done by appending the two branches on the previous command:
+```
+nx affected:test --base=origin/master --head=your-branch
+```
+
+## Restrictions
+You may want to categorize libraries in different domains and limit the dependencies between those domains to improve maintainability.
+Nx can help you with this, by restricting access between domains or by allowing the dependency flow to go only one way around.
+To configure and manage these restrictions, you can apply tags in the nx.json file for apps and libraries.
```
Comment (orjandesmet) [inline _posts/2019-12-02-Nx.md]: You also have to setup the linting rules in tslint.json ;)

File: _posts/2019-12-02-Nx.md [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+These libraries will expose the code using a public API.
+
+The code inside these libraries can be imported afterwards in your apps, without having to fetch them from a npm registry.
+Think about how easily you can share an interface between your frontend and backend, without having to break it on one side.
+Whenever something changes in the API, both sides will get updated so they will both remain in sync.
+
+# Conclusion
+Having worked within a monorepo myself for the last months, I really appreciate the way it improved the daily workflow.
+It encourages me to write more reusable code and to keep features small.
+Other teams may have the same requirements, so code can be easily migrated to libs.
```
Comment (orjandesmet) [inline _posts/2019-12-02-Nx.md]: Though beware to not create a giant "shared" lib. Any change to that will affect a lot of applications. It's better to have more small libs than less fat libs.

File: _posts/2019-12-02-Nx.md [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+The complexity will increase and after a while the whole platform becomes unmanageable.
+
+# Monorepo
+
+The problems mentioned above led large software companies to transition the management of their separate projects by bringing them together under one repository: the monorepo.
+Inside the monorepo, you have all the applications working together while you’re developing your app.
+You’ll get feedback immediately when the service you’re calling returns a different answer than expected or when the page you are navigating to is located on another path.
+When you're not sure how a service of the platform is working, you can just hop in the source code inside the repo to find out more about it.
+Code sharing becomes a breeze, because all the code is available in one repository.
+You don’t have to deal anymore with npm packages for shared code in your organization or with the separate configurations, pipelines and versions of the shared libs.
```
Comment (orjandesmet) [inline _posts/2019-12-02-Nx.md]: Because there is only 1 package.json used for all projects.

File: _posts/2019-12-02-Nx.md [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+## Building
+When new code is introduced or something has been refactored, Nx will be able to find out which dependencies are affected by your changes.
+This gives you the opportunity to only build the affected dependencies and only run the tests of those affected ones.
+For your build pipeline, that means that testing the affected apps can greatly improve the build time when dealing with large workspaces.
+
+For example, if we edit something in the shop app, our dependency graph will look like this:
+
+<img src="{{ '/img/2019-12-02-nx/dependency_graph_affected.png' | prepend: site.baseurl }}" alt="Dependency graph of affected apps" class="image fit" style="margin:0px auto; max-width: 500px;">
+
+Targetting the affected code can be done using the affected command.
```
Comment (yannickdeturck) [inline _posts/2019-12-02-Nx.md]:
```suggestion
Targeting the affected code can be done using the affected command.
```

File: _posts/2019-12-02-Nx.md [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+The unit and end-to-end tests are both configured for the shop app that we've just created. They can be run by using `nx test shop` and `nx e2e shop`, respectively.
+If you want to learn more about Jest or Cypress, be sure to check out the awesome blog posts that my colleagues have wrote:
+
+[Testing Angulat with Jest](https://ordina-jworks.github.io/testing/2018/08/03/testing-angular-with-jest.html){:target="_blank" rel="noopener noreferrer"}
+
+[Cypress: a new kid on the E2E block](https://ordina-jworks.github.io/testing/2019/07/18/Cypress.html){:target="_blank" rel="noopener noreferrer"}
+
+## Code formatting
+
+<img src="{{ '/img/2019-12-02-Nx/prettier.png' | prepend: site.baseurl }}" alt="Prettier" class="image right">
```
[the remark this answers is not in the document]
Comment (DimiDeKerf) [reply _posts/2019-12-02-Nx.md]: Yeah seems like the image wasn't included in the commit...

File: _posts/2019-12-02-Nx.md [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+The code inside these libraries can be imported afterwards in your apps, without having to fetch them from a npm registry.
+Think about how easily you can share an interface between your frontend and backend, without having to break it on one side.
+Whenever something changes in the API, both sides will get updated so they will both remain in sync.
+
+# Conclusion
+Having worked within a monorepo myself for the last months, I really appreciate the way it improved the daily workflow.
+It encourages me to write more reusable code and to keep features small.
+Other teams may have the same requirements, so code can be easily migrated to libs.
+Switching between features is much easier and accessible, since I don’t have to check out another repository or getting familiar with the way of working of another team.
+Overall, I feel more connected with other developers and with the platform while building apps in the same repository.
```
[the remark this answers is not in the document]
Comment (DimiDeKerf) [reply _posts/2019-12-02-Nx.md]: For now I haven't encountered any yet 😄

**Comment by DimiDeKerf** [comment]
Thanks for reviewing @yannickdeturck and @orjandesmet !

**Pull request merged**